### PR TITLE
List all anomaly tests on the anomalies page.

### PR DESCRIPTION
**Pull request by lukebjerring**
Running at https://lukebjerring-interoperability-anomalies-dot-wptdashboard.appspot.com/interop/anomalies

First iteration is a slow-loading, flat list of all tests which pass in some, but not all, browsers, with no browser-specific views.

`/interop/anomalies` (no params) will show a list of anomalies (test that fail in at least one, but not all, browsers).
`/interop/anomalies?browser=[browser]` will show tests that **fail** in that browser, but not in all browsers.

### CODE REVIEW

**Comment by rwaldron** [comment]
> This PR aggregates several other pending PRs

Why? Those PRs were just in review. Two of them were ready to land and are now (#334, #338), one of them landed immediately because it was so small (#348) and #347 is in normal review (and likely to land shortly). I don't understand the benefit in "aggregating" open PRs that are each in their own review cycle. What happens when feedback on one of the PRs results in changes on that branch? Are you manually maintaining those changes across branches? I'm not sure what to do with this now ¯\_(ツ)_/¯

**Comment by lukebjerring** [comment]
> Why? [...]

Because there is no way to do child-PR in GitHub based on a pending PR from a forked repo.
Once the other PRs have landed, the diff of this PR becomes the actual changes, but I'm forced to include all changes in a single PR, so I commented on the aggregation to help indicate where many of the (repeated) changes come from.

**Comment by lukebjerring** [comment]
This PR highlights the following issues:

- There's a bug in the FailureMetadata export where all test dirs and names are the same.
- Sorting the metadata with a comparator client-side is extremely slow (@43k entries, taking 20s)
- The metadata files are huge (>20M)

**Comment by rwaldron** [comment]
@lukebjerring given the issues you've discovered, is this PR still ready for landing, or does it require more work?

**Comment by lukebjerring** [comment]
IMO, the work needed to correct the issues is tangential to this PR itself, so I think it deserves review as-is.

File: webapp/components/anomalies.html [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        let anomalies = metrics.data;
+        if (!this.browser) {
+          anomalies = anomalies.filter(d => this._isTestFile(d.dir));
+        }
+        
+        const totalBrowsers = metrics.metadata.test_runs.length;
+        let matchesPassRate = !this.browser
+          ? data => data.pass_rates[passes] < 1 || !this._isTestFile(data.dir)
+          : data => data.num_other_failures + 1 < totalBrowsers;
+        let testPath = !this.browser
```
Comment (rwaldron) [inline webapp/components/anomalies.html]: `testPath` is unused. On line 98, `let testPath = ...` occurs and shadows this binding, but _this_ particular `testPath` is not used before that point.

File: webapp/components/anomalies.html [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        const metrics = await this._fetchResults(this.passRateMetadata.url);
+
+        let anomalies = metrics.data;
+        if (!this.browser) {
+          anomalies = anomalies.filter(d => this._isTestFile(d.dir));
+        }
+        
+        const totalBrowsers = metrics.metadata.test_runs.length;
+        let matchesPassRate = !this.browser
+          ? data => data.pass_rates[passes] < 1 || !this._isTestFile(data.dir)
```
Comment (rwaldron) [inline webapp/components/anomalies.html]: `passes` is not defined.

**Comment by rwaldron** [comment]
@lukebjerring first review pass complete.

File: webapp/components/anomalies.html [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+<link rel="import" href="test-run.html">
+
+<dom-module id="wpt-anomalies">
+  <template>
+    <style>
+    </style>
+
+      <section class="caveat">
+        Data below are intended for web platform implementers and do not contain useful metrics for evaluation or comparison of web platform features. Also note that tested Edge and Safari are not pre-release versions (<a href="https://github.com/w3c/wptdashboard/issues/109">#109</a>, <a href="https://github.com/w3c/wptdashboard/issues/110">#110</a>).
+      </section>
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        // Fetch the results JSON for the TestRuns
+        const metrics = await this._fetchResults(this.passRateMetadata.url);
+
+        let anomalies = metrics.data;
+        if (!this.browser) {
+          anomalies = anomalies.filter(d => this._isTestFile(d.dir));
+        }
+        
+        const totalBrowsers = metrics.metadata.test_runs.length;
+        let matchesPassRate = !this.browser
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html (hunk 2 above)
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html (hunk 1 above)
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          : data => data.num_other_failures + 1 < totalBrowsers;
+        let testPath = !this.browser
+          ? data => data.dir
+          : data => data.test.test
+
+        // For each pass-rate (browser count), build a tree of tests + directories.
+        let byPassRate = [];
+        // We only want anomalies; at least one pass, at most N-1 passes.
+        for (let passes = 1; passes < totalBrowsers; passes++) {
+          byPassRate.push({ 'passingBrowsers': passes, 'tests': [] });
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            for (let i in data.pass_rates) {
+              const pass_rate = data.pass_rates[i];
+              if (i < 1 || i >= totalBrowsers || pass_rate < 1) {
+                continue
+              }
+              byPassRate[i - 1].tests.push(testPath);
+            }
+          }
+        }
+        this.anomalyGroups = byPassRate
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      }
+
+      async _fetchResults (url) {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        const response = await window.fetch(url);
+        return response.ok ? await response.json() : {};
+      }
+
+      _isTestFile(path) {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done (moved outside class decl)

File: webapp/components/anomalies.html [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        return path.endsWith('.html') || path.endsWith('.htm') ||
+               path.endsWith('.py') || path.endsWith('.svg') ||
+               path.endsWith('.xhtml') || path.endsWith('.xht') ||
+               path.endsWith('.xml');
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

**Comment by lukebjerring** [comment]
Lost your Review context when I rebase anyway, so changes (including review changes) are squashed in the one commit.
Re-deployed to https://lukebjerring-interoperability-anomalies-dot-wptdashboard.appspot.com/interop/anomalies (using `git/git-deploy` :smiling_imp:)

File: webapp/templates/anomalies.html [hunk 10]
```diff
@@ -1,11 +1,14 @@
...
+<head>
+  {{ template "_head_common.html" }}
+  <link rel="import" href="/components/anomalies.html">
+</head>
 <body>
 <div id="content">
   {{ template "_header.html" }}
   <div>
-    Interoperability anomalies view coming soon.
+    <wpt-anomalies pass-rate-metadata="{{ .Metadata }}" browser="{{ .Browser }}"></wpt-anomalies>
```
Comment (rwaldron) [inline webapp/templates/anomalies.html]: When I view source on the link you provided, I'm seeing `browser=""` (but plenty of `pass-rate-metadata="..."`). Is that expected?

Comment (lukebjerring) [reply webapp/templates/anomalies.html]: If you set ?browser=chrome, it will set the browser-specific view up.

File: webapp/components/anomalies.html [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    <template is="dom-repeat" items="{{anomalyGroups}}" as="anomalyGroup" itemsIndexAs="index">
+      <template is="dom-if" if="{{browser}}">
+        <h2>Failing in [[browser]] and [[index]] other browsers</h2>
+      </template>
+      <template is="dom-if" if="{{!browser}}">
+        <h2>Passing in [[anomalyGroup.passingBrowsers]] browsers</h2>
+      </template>
+      <template is="dom-repeat" items="{{anomalyGroup.tests}}" as="test">
+        <a href="[[test]]">[[test]]</a><br />
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

File: webapp/components/anomalies.html [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    <section class="caveat">
+      Data below are intended for web platform implementers and do not contain useful metrics for evaluation or comparison of web platform features. Also note that tested Edge and Safari are not pre-release versions (<a href="https://github.com/w3c/wptdashboard/issues/109">#109</a>, <a href="https://github.com/w3c/wptdashboard/issues/110">#110</a>).
+    </section>
+
+    <template is="dom-repeat" items="{{anomalyGroups}}" as="anomalyGroup" itemsIndexAs="index">
+      <template is="dom-if" if="{{browser}}">
+        <h2>Failing in [[browser]] and [[index]] other browsers</h2>
+      </template>
+      <template is="dom-if" if="{{!browser}}">
+        <h2>Passing in [[anomalyGroup.passingBrowsers]] browsers</h2>
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Done.

**Comment by lukebjerring** [comment]
Changes pushed + redeployed to https://lukebjerring-interoperability-anomalies-dot-wptdashboard.appspot.com/interop/anomalies. PTAL.

File: webapp/templates/anomalies.html (hunk 10 above)
Comment (rwaldron) [reply webapp/templates/anomalies.html]: > If you set ?browser=chrome, it will set the browser-specific view up.

When submitting PRs for review, please include as much information as possible. Thanks.

File: webapp/components/anomalies.html [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    <section class="caveat">
+      Data below are intended for web platform implementers and do not contain useful metrics for evaluation or comparison of web platform features. Also note that tested Edge and Safari are not pre-release versions (<a href="https://github.com/w3c/wptdashboard/issues/109">#109</a>, <a href="https://github.com/w3c/wptdashboard/issues/110">#110</a>).
+    </section>
+
+    <template is="dom-repeat" items="{{anomalyGroups}}" as="anomalyGroup" itemsIndexAs="index">
+      <template is="dom-if" if="{{browser}}">
+        <h2>Failing in [[browser]] and [[computeOtherBrowserCount(index)]]</h2>
+      </template>
+      <template is="dom-if" if="{{!browser}}">
+        <h2>Passing in [[computeBrowserCount(anomalyGroup.passingBrowsers)]]</h2>
```
Comment (rwaldron) [inline webapp/components/anomalies.html]: I would've deferred pluralization semantics to a third party library. That can be done in a follow up

**Comment by rwaldron** [comment]
Similar to the interop branch, I can't test this locally because the go script for making local metrics data available is not functioning correctly. Large additions like this and the other need to be fully documented with "how to stand this up" instructions. Opening big PRs with a link to an ad-hoc staging URL is not sufficient for review—we need to be able to reproduce the entire environment to understand the full scope of the new and modified application code.

**Comment by lukebjerring** [comment]
I'm confused - this PR does not modify the local setup. 'local repro' is for development (populate-dev-data), and testing local repro would only confirm that the local dev flow is working. When sending you the ad-hoc staging environments, I am pushing to the AppEngine environment using the standard `--no-promote` deploy, which is, by far, a better test of the state of the repo for the PR (and IMO a much more important test for ensuring production deployment will be successful).

I've opened an issue for `populate_dev_data.go` not working; could you expand on the problem? https://github.com/w3c/wptdashboard/issues/378

File: webapp/components/anomalies.html (hunk 13 above)
Comment (lukebjerring) [reply webapp/components/anomalies.html]: Opened https://github.com/w3c/wptdashboard/issues/379 to track this. Did you have a specific library in mind?

**Comment by rwaldron** [comment]
> I'm confused - this PR does not modify the local setup. 'local repro' is for development (populate-dev-data), and testing local repro would only confirm that the local dev flow is working

Being able to run wpt.fyi locally, make changes, poke and prod the application code is a hard dependency. I'm unable to run this locally because populate-dev-data isn't working :P

**Comment by rwaldron** [comment]
Whoops... forgot this...

> Edit: Note also that the (static) data locally is a copy that predates #356 being resolved. I've opened #382 to update to more recent run.

Excellent!

**Comment by lukebjerring** [comment]
Provided #378 was actually fixed via #381 for you, this should be ready for your eyeballs.

**Comment by rwaldron** [comment]
Awesome, my local dev looks just like your staging! Thanks for doing the follow up on that.
